src/Eloquent/Traits/HasAdjacencyList.php

```diff
@@ -159,15 +159,16 @@ public function ancestorsAndSelf()
     /**
      * Instantiate a new Ancestors relationship.
      *
-     * @param \Illuminate\Database\Eloquent\Builder $query
-     * @param \Illuminate\Database\Eloquent\Model $parent
+     * @param \Illuminate\Database\Eloquent\Builder<self> $query
+     * @param self $parent
      * @param string $foreignKey
      * @param string $localKey
      * @param bool $andSelf
      * @return \Staudenmeir\LaravelAdjacencyList\Eloquent\Relations\Ancestors<self>
      */
     protected function newAncestors(Builder $query, Model $parent, $foreignKey, $localKey, $andSelf)
     {
+        // @phpstan-ignore return.type
```

Comment:
the generics for these custom relations are really difficult, not sure how to remove these exclusions, I'm probably missing some generic somewhere but I can't find it

Reply:
These don't matter so much, what really matters is the method signature

Reply:
I noticed that it works for methods like `newBloodline()` and `newRootAncestor()`, but I don't see how they are different from `newAncestors()` etc.

Reply:
Same, I miss some in-depth knowledge of this area of PHPStan to know the fine details of what is different / what's causing the error in this instance

Reply:
@calebdw Do you have any idea here? The methods `newAncestors()` and `newRootAncestor()` and their respective classes `Ancestors`/`RootAncestor` have the same structure & types (to me), but get analyzed differently for some reason.

Reply:
Sorry, I haven't had a chance to dig into this yet, but I try to get to it over the weekend

Reply:
@staudenmeir,

Traits are analyzed twice: by themselves and then in the *context* of where they are used. I suspect that the `newAncestors` method was being used in the context of a Model whereas the `newRootAncestor` was not and that was causing the error to be thrown.

However, after removing the `tests` path from the phpstan analysis locally this no longer errors.